File: CS_Algorithms/ADA_Algorithms/Greedy_Algorithms/C++/fractional_knapsack.cpp
#include<iostream>
#include<iomanip>
#include<algorithm>
#include<vector>
using namespace std;

class Item {
public:
    string index; double value; double weight;
    Item(const string&idx,double wt,double val):index(idx),weight(wt),value(val){}
    double ratio() const{ return value/weight; }
};

class FractionalKnapsack {
public:
    double knapsack_capacity; vector<Item>items; 
    FractionalKnapsack(double knapsack_capacity, const vector<Item>& items): knapsack_capacity(knapsack_capacity), items(items) {}
    void maximize_profit(){
        sort(items.begin(),items.end(),[](const Item&a,const Item&b){ return a.ratio()>b.ratio(); });
        double total_value = 0; double remaining_capacity = knapsack_capacity;
        for(const auto& item: items){
            if(remaining_capacity<=0) break;
            if(item.weight<=remaining_capacity){
                total_value += item.value;
                remaining_capacity -= item.weight;
                cout << "Taking full item: " << item.index << " (Value: " << item.value << ", Weight: " << item.weight << ")\n";
            } else {
                double fraction = remaining_capacity / item.weight;
                total_value += item.value * fraction;
                cout << "Taking fraction of item: " << item.index << " (Value: " << item.value * fraction << ", Weight: " << remaining_capacity << ")\n";
                remaining_capacity = 0; 
            }  
    }
    cout<< total_value << endl;
}
};

int main(){
    vector<Item>items={Item("A",5,50),Item("B",15,80),Item("C",35,90)};
    FractionalKnapsack fk(45, items);
    fk.maximize_profit();
}

File: CS_Algorithms/ADA_Algorithms/Greedy_Algorithms/Python/fractional_knapsack.py
from typing import List 

class Item:
    def __init__(self,index:int,value:float,weight:float):
        self.index = index
        self.value = value
        self.weight = weight
    @property
    def ratio(self):
        return self.value / self.weight
    
class FractionalKnapsack:

    def __init__(self,capacity:float, items:List[Item]):
        self.capacity = capacity
        self.items = items
    def maximize_profit(self):
        sorted_items = sorted(self.items, key=lambda x: x.ratio, reverse=True)
        remaining_capacity = self.capacity
        total_profit = 0.0
        selected_items = []
        for item in sorted_items:
            if remaining_capacity <= 0:
                break
            if item.weight <= remaining_capacity:
                selected_items.append((item.index,1.0,item.value))
                total_profit += item.value
                remaining_capacity -= item.weight
            else:
                fraction = remaining_capacity / item.weight
                selected_items.append((item.index,fraction,fraction*item.value))
                total_profit += fraction * item.value
                remaining_capacity = 0
        return total_profit, selected_items
    
def main():
    items=[Item(1,5,50),Item(2,15,80),Item(3,35,90)]
    capacity=45
    profit, selected_items = FractionalKnapsack(capacity, items).maximize_profit()
    print(f"Total Profit: {profit}")
    print("Selected Items:")
    for index, fraction, item_profit in selected_items:
        print(f"Item {index}: {fraction:.2f} (Profit: {item_profit:.2f})")  

if __name__ == "__main__":
    main()
